File: source/allCategories.m
function categories = allCategories(varargin)
%ALLCATEGORIES Set or get all the categories on which we are working.
%
% Syntax:
%   V = allCategories()
%       Get the set of all the categories on which we are working.
%
%   allCategories(V)
%       Set the categories on which we are working to V.

% Copyright 2023 Acadèmia de su Sardu APS
persistent storedCategories

assert(nargin>=0 & nargin<=1, ...
    "This function can only be called with zero or one argument.");

if nargin==0
    categories = storedCategories;
    assert(~isempty(categories), ...
        "You should first set the categories by calling 'allCategories' with an input");
end
if nargin==1
    argument = varargin{1};
    assert( ...
        (isstring(argument) | iscellstr(argument)) & length(argument)>1, ...
        "Input should be an array of strings or a cell array containing chars with more than one element.");
    storedCategories = unique(string(argument));
end
end

File: python/linguistic_variation_toolbox/all_categories.py
from __future__ import annotations

_stored_categories: list[str] | None = None


def all_categories(categories: list[str] | None = None) -> list[str] | None:
    """Get or set the global list of linguistic categories.

    Call with no argument to get; call with a list to set.
    The list must have more than one element.
    """
    global _stored_categories
    if categories is None:
        assert _stored_categories is not None, (
            "You should first set the categories by calling 'all_categories' with an input"
        )
        return _stored_categories
    categories = list(categories)
    assert len(categories) > 1, (
        "Input should be a list of strings with more than one element."
    )
    _stored_categories = sorted(set(str(c) for c in categories))
    return None
